File: python/aip_token/policy.py
"""Simple policy profiles for Biscuit-based AIP tokens."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone


@dataclass
class SimplePolicy:
    """A simple policy that generates Datalog checks for Biscuit tokens.

    Budget is expressed in integer cents (Biscuit has no float support).
    """

    tools: list[str] = field(default_factory=list)
    budget_cents: int | None = None
    max_depth: int | None = None
    ttl_seconds: int | None = None

    def to_datalog(self) -> str:
        rules: list[str] = []
        if self.tools:
            tool_list = ", ".join(f'"{t}"' for t in self.tools)
            rules.append(f"check if tool($tool), [{tool_list}].contains($tool);")
        if self.budget_cents is not None:
            rules.append(f"check if budget($b), $b <= {self.budget_cents};")
        if self.max_depth is not None:
            rules.append(f"check if depth($d), $d <= {self.max_depth};")
        if self.ttl_seconds is not None:
            expiry = datetime.now(tz=timezone.utc) + timedelta(seconds=self.ttl_seconds)
            rules.append(
                f'check if time($t), $t <= {expiry.strftime("%Y-%m-%dT%H:%M:%SZ")};'
            )
        return "\n".join(rules)

File: rust/aip-token/src/policy.rs
use std::time::{Duration, SystemTime};
use chrono::{DateTime, Utc};

/// Simple policy profile -- generates canonical Datalog from user-specified values.
/// Budget is in integer cents (Biscuit Datalog has no float type).
pub struct SimplePolicy {
    pub tools: Vec<String>,
    pub budget_cents: Option<i64>,
    pub max_depth: Option<u32>,
    pub ttl_seconds: Option<u64>,
}

/// Policy profile enum for all supported profiles.
pub enum PolicyProfile {
    Simple(SimplePolicy),
    Standard(String),   // raw Datalog string
    Advanced(String),   // raw Datalog string
}

impl SimplePolicy {
    /// Generate canonical Datalog. These templates are normative per the AIP spec.
    /// Implementations MUST generate exactly these patterns.
    pub fn to_datalog(&self) -> String {
        let mut rules = Vec::new();

        if !self.tools.is_empty() {
            let tool_list = self.tools.iter()
                .map(|t| format!("\"{}\"", t))
                .collect::<Vec<_>>()
                .join(", ");
            rules.push(format!("check if tool($tool), [{}].contains($tool);", tool_list));
        }

        if let Some(cents) = self.budget_cents {
            rules.push(format!("check if budget($b), $b <= {};", cents));
        }

        if let Some(depth) = self.max_depth {
            rules.push(format!("check if depth($d), $d <= {};", depth));
        }

        if let Some(ttl) = self.ttl_seconds {
            let expiry: DateTime<Utc> = (SystemTime::now() + Duration::from_secs(ttl)).into();
            rules.push(format!("check if time($t), $t <= {};", expiry.format("%Y-%m-%dT%H:%M:%SZ")));
        }

        rules.join("\n")
    }
}

impl PolicyProfile {
    pub fn to_datalog(&self) -> String {
        match self {
            Self::Simple(p) => p.to_datalog(),
            Self::Standard(d) | Self::Advanced(d) => d.clone(),
        }
    }
}
